modules/l4postgres/matcher.go

```diff
@@ -0,0 +1,133 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package l4postgres allows the L4 multiplexing of Postgres connections
+//
+// With thanks to docs and code published at these links:
+// ref: https://github.com/mholt/caddy-l4/blob/master/modules/l4ssh/matcher.go
+// ref: https://github.com/rueian/pgbroker/blob/master/message/startup_message.go
+// ref: https://github.com/traefik/traefik/blob/master/pkg/server/router/tcp/postgres.go
+// ref: https://ivdl.co.za/2024/03/02/pretending-to-be-postgresql-part-one-1/
+// ref: https://www.postgresql.org/docs/current/protocol-message-formats.html#PROTOCOL-MESSAGE-FORMATS-STARTUPMESSAGE
+// ref: https://www.postgresql.org/docs/current/protocol-message-formats.html#PROTOCOL-MESSAGE-FORMATS-SSLREQUEST
+package l4postgres
+
+import (
+	"encoding/binary"
+	"errors"
+	"io"
+
+	"github.com/caddyserver/caddy/v2"
+	"github.com/mholt/caddy-l4/layer4"
+)
+
+func init() {
+	caddy.RegisterModule(MatchPostgres{})
+}
+
+const (
+	// Magic number to identify a SSLRequest message
+	sslRequestCode = 80877103
+	// byte size of the message length field
+	initMessageSizeLength = 4
+)
+
+// Message provides readers for various types and
+// updates the offset after each read
+type message struct {
+	data   []byte
+	offset uint32
+}
+
+func (b *message) ReadUint32() (r uint32) {
+	r = binary.BigEndian.Uint32(b.data[b.offset : b.offset+4])
+	b.offset += 4
+	return r
+}
+
+func (b *message) ReadString() (r string) {
+	end := b.offset
+	max := uint32(len(b.data))
+	for ; end != max && b.data[end] != 0; end++ {
+	}
+	r = string(b.data[b.offset:end])
+	b.offset = end + 1
+	return r
+}
+
+// NewMessageFromBytes wraps the raw bytes of a message to enable processing
+func newMessageFromBytes(b []byte) *message {
+	return &message{data: b}
+}
+
+// StartupMessage contains the values parsed from the startup message
+type startupMessage struct {
+	ProtocolVersion uint32
+	Parameters      map[string]string
+}
+
+// MatchPostgres is able to match Postgres connections.
+type MatchPostgres struct{}
+
+// CaddyModule returns the Caddy module information.
+func (MatchPostgres) CaddyModule() caddy.ModuleInfo {
+	return caddy.ModuleInfo{
+		ID:  "layer4.matchers.postgres",
+		New: func() caddy.Module { return new(MatchPostgres) },
+	}
+}
+
+// Match returns true if the connection looks like the Postgres protocol.
+func (m MatchPostgres) Match(cx *layer4.Connection) (bool, error) {
+	// Get bytes containing the message length
+	head := make([]byte, initMessageSizeLength)
+	if _, err := io.ReadFull(cx, head); err != nil {
+		return false, err
+	}
+
+	// Get actual message length
+	data := make([]byte, binary.BigEndian.Uint32(head)-initMessageSizeLength)
+	if _, err := io.ReadFull(cx, data); err != nil {
+		return false, err
+	}
+
+	b := newMessageFromBytes(data)
+
+	// Check if it is a SSLRequest
+	code := b.ReadUint32()
+	if code == sslRequestCode {
```

Comment:
I think `cx.Write('S')` would go here, but then how to apply a TLS matcher?

Reply:
A matcher is the wrong place to write outputs. Writings go into handlers.

Reply:
How do you propose to finish matching and pass to a handler if you don't have the bytes to match yet?

Or are you suggesting that the rest of the matching be completed in a handler?

Reply:
There would be another set of matcher/handler chain, akin to the [subroute handler](https://caddyserver.com/docs/json/apps/http/servers/routes/handle/subroute/) that's available in the `http` app. Assuming the existence of a `subroute` handler and a `writer` handler for the convenience of the discussion, it'll look something like this:

```json
{
	"routes": [
		{
			"match": [
				{
					"postgres": {}
				}
			],
			"handle": [
				{
					"handler": "writer",
					"bytes": "S"
				},
				{
					"handler": "subroute",
					"routes": [
						{
							"match": [
								{
									"tls": {
										"sni": [
											"postgres.example.net"
										]
									}
								}
							],
							"handle": [
								{
									"handle": "tls"
								},
								{
									"handle": "proxy",
									"upstreams": [
										{
											"dial": "postgres.localhost:5432"
										}
									]
								}
							]
						}
					]
				}
			]
		}
	]
}
```

Reply:
The abstraction has to be a bit higher than that, otherwise what you'll end up with is a very tedious programming language for which the AST is literally this JSON structure of matchers and handlers.

That said, I kinda get what you're driving at.

To convey the idea without writing out all the syntax:

```
matcher: postgres
  - handler: postgres-ssl
    # encrypted connection
    - matcher: tls
      - sni: customer1.example.com
        - handler: postgres-tls-proxy
    # plain connection
    - matcher: postgres-connection
      - hostname: customer1.example.com
        - handler: proxy-plain-proxy
```

You couldn't use the normal `proxy` because you have to replay the SSLRequest ``, eat the `S`, and then resume the connection.

This is so highly specific to postgres... just seems like it'll be a real pain to break it down into generic abstractions just to try to fit into a pattern that was created without the foresight of this type of interaction.

Reply:
Thinking on it a bit more: This is very much *NOT* a generic flow. There are a finite number of outcomes:
1. Plain Postgres TCP (`sslmode=disable`) - match on DB and proxy
2. SSLRequest (`sslmode=default|require|verify-partial|verify-full`) \
   no matching info available yet, so the config must indicate whether to send `S` to the client... or whatever byte means "not encrypted"
   - Proceed to 1 (above) if the config is not encrypted
   - The connection is encrypted, so there are now 2 options:
       1. terminate TLS now, then proceed to 1 (above)
       2. match on TLS, then replay the client connection, then proxy

Reply:
At this point, you should move this conversation and conclusion to #187. @metafeather may have better reasoning over this than I because he's working with this.